components/blitz/src/omero/gateway/facility/DataManagerFacility.java

```diff
@@ -497,4 +572,242 @@ public void addImagesToDataset(SecurityContext ctx,
+                    long pos = 0;
+                    int rlen;
+                    byte[] buf = new byte[INC];
+                    ByteBuffer bbuf;
+                    while (((rlen = stream.read(buf)) > 0) && !cb.isCancelled()) {
+                        rawFileStore.write(buf, pos, rlen);
+                        pos += rlen;
+                        bbuf = ByteBuffer.wrap(buf);
+                        bbuf.limit(rlen);
+                    }
+                    stream.close();
+
+                    if (cb.isCancelled()) {
+                        try {
+                            rawFileStore.close();
+                        } catch (ServerError e) {
+                        }
+                        cb.setResult(null);
+                        return;
+                    }
+
+                    originalFile = rawFileStore.save();
+
+                    FileAnnotation fa = new FileAnnotationI();
+                    fa.setFile(originalFile);
+                    if (description != null)
+                        fa.setDescription(omero.rtypes.rstring(description));
+                    fa.setNs(omero.rtypes.rstring(namespace));
+                    fa = (FileAnnotation) saveAndReturnObject(ctx, fa);
+
+                    if (target != null)
+                        cb.setResult(attachAnnotation(ctx,
+                                new FileAnnotationData(fa), target));
+                    else
+                        cb.setResult(new FileAnnotationData(fa));
+                } catch (Throwable t) {
+                    cb.setException(t);
+                } finally {
+                    if (rawFileStore != null) {
+                        try {
+                            rawFileStore.close();
+                        } catch (ServerError e) {
+                        }
+                    }
+                }
+            }
+        };
+
+        if (callback != null) {
+            (new Thread(r)).start();
+            return null;
+        }
+        
+        r.run();
+        return (FileAnnotationData) cb.result;
+    }
+    
+    /**
+     * Create/attach an {@link AnnotationData} to a given {@link DataObject}
+     * 
+     * @param ctx
+     *            The {@link SecurityContext}
+     * @param annotation
+     *            The {@link AnnotationData}
+     * @param target
+     *            The {@link DataObject} to attach to
+     * @return The {@link AnnotationData}
+     * @throws DSOutOfServiceException
+     *             If the connection is broken, or not logged in
+     * @throws DSAccessException
+     *             If an error occurred while trying to retrieve data from OMERO
+     *             service.
+     */
+    public <T extends AnnotationData> T attachAnnotation(SecurityContext ctx,
+            T annotation, DataObject target) throws DSOutOfServiceException,
+            DSAccessException {
+        if (target != null) {
+            if (target instanceof ProjectData) {
```

Comment:
As a longer-term RFE, it'd be nice if we could reduce the use of `instanceof` in favor of interface-based methods on the `*Data` objects themselves.